lib/test.js

```diff
@@ -104,51 +112,52 @@ Test.prototype.plan = function (count) {
 };
 
 Test.prototype.run = function () {
-	this.promise = {};
+	var self = this;
+
+	this.promise = Promise.pending();
 
 	// TODO(vdemedes): refactor this to avoid storing the promise
-	return new Promise(function (resolve, reject) {
-		this.promise.resolve = resolve;
-		this.promise.reject = reject;
+	if (!this.fn) {
+		return this.exit();
+	}
 
-		if (!this.fn) {
-			return this.exit();
-		}
+	this._timeStart = Date.now();
 
-		this._timeStart = Date.now();
+	// wait until all assertions are complete
+	this._timeout = setTimeout(noop, maxTimeout);
 
-		// wait until all assertions are complete
-		this._timeout = setTimeout(function () {}, maxTimeout);
+	var ret;
 
-		try {
-			var ret = this.fn(this);
+	try {
+		ret = this.fn(this);
+	} catch (err) {
+		this.assertError = err;
+		this.exit();
+	}
 
-			if (ret && ret[observableSymbol]) {
-				ret = ret[observableSymbol]().forEach(noop);
-			}
+	if (isObservable(ret)) {
+		ret = ret[observableSymbol]().forEach(noop);
+	}
 
-			if (ret && typeof ret.then === 'function') {
-				ret
-					.then(function () {
-						if (!this.planCount || this.planCount === this.assertCount) {
-							this.exit();
-						}
-					}.bind(this))
-					.catch(function (err) {
-						this.assertError = new assert.AssertionError({
-							actual: err,
-							message: 'Promise rejected → ' + err,
-							operator: 'promise'
-						});
-
-						this.exit();
-					}.bind(this));
-			}
-		} catch (err) {
-			this.assertError = err;
-			this.exit();
-		}
-	}.bind(this));
+	if (isThenable(ret)) {
+		ret
+			.then(function () {
+				if (!self.planCount || self.planCount === self.assertCount) {
+					self.exit();
+				}
+			})
+			.catch(function (err) {
+				self.assertError = new assert.AssertionError({
+					actual: err,
+					message: 'Promise rejected → ' + err,
+					operator: 'promise'
+				});
+
+				self.exit();
+			});
+	}
+
+	return this.promise.promise;
```

Comment:
because `Promise.pending()` returns something called a `resolver`, it's not an actual promise, it's not thenable. To get a promise from resolver, we need to get it from `promise` property:

```js
var resolver = Promise.pending();
var promise = resolver.promise;

promise.then(...);
```